crates/loro-internal/src/state/container_store/container_wrapper.rs

```diff
@@ -0,0 +1,279 @@
+            depth,
+            parent,
+            kind: idx.get_type(),
+            state: Some(state),
+            bytes: None,
+            value: None,
+            bytes_offset_for_state: None,
+            bytes_offset_for_value: None,
+            flushed: false,
+        }
+    }
+
+    pub fn depth(&self) -> usize {
+        self.depth
+    }
+
+    /// It will not decode the state if it is not decoded
+    #[allow(unused)]
+    pub fn try_get_state(&self) -> Option<&State> {
+        self.state.as_ref()
+    }
+
+    /// It will decode the state if it is not decoded
+    pub fn get_state(&mut self, idx: ContainerIdx, ctx: ContainerCreationContext) -> &State {
+        self.decode_state(idx, ctx).unwrap();
+        self.state.as_ref().expect("ContainerWrapper is empty")
+    }
+
+    /// It will decode the state if it is not decoded
+    pub fn get_state_mut(
+        &mut self,
+        idx: ContainerIdx,
+        ctx: ContainerCreationContext,
+    ) -> &mut State {
+        self.decode_state(idx, ctx).unwrap();
+        self.bytes = None;
+        self.value = None;
+        self.flushed = false;
+        self.state.as_mut().unwrap()
+    }
+
+    pub fn get_value(&mut self, idx: ContainerIdx, ctx: ContainerCreationContext) -> LoroValue {
+        if let Some(v) = self.value.as_ref() {
+            return v.clone();
+        }
+
+        self.decode_value(idx, ctx).unwrap();
+        if self.value.is_none() {
+            return self.state.as_mut().unwrap().get_value();
+        }
+
+        self.value.as_ref().unwrap().clone()
+    }
+
+    pub fn encode(&mut self) -> Bytes {
+        if let Some(bytes) = self.bytes.as_ref() {
+            return bytes.clone();
+        }
+
+        // ContainerType
+        // Depth
+        // ParentID
+        // StateSnapshot
+        let mut output = Vec::new();
+        output.push(self.kind.to_u8());
+        leb128::write::unsigned(&mut output, self.depth as u64).unwrap();
+        postcard::to_io(&self.parent, &mut output).unwrap();
+        self.state
+            .as_mut()
+            .unwrap()
+            .encode_snapshot_fast(&mut output);
+        let ans: Bytes = output.into();
+        self.bytes = Some(ans.clone());
+        ans
+    }
+
+    pub fn decode_parent(b: &[u8]) -> Option<ContainerID> {
+        let mut bytes = &b[1..];
+        let _depth = leb128::read::unsigned(&mut bytes).unwrap();
+        let (parent, _bytes) = postcard::take_from_bytes::<Option<ContainerID>>(bytes).unwrap();
+        parent
+    }
+
+    pub fn new_from_bytes(b: Bytes) -> Self {
+        let src: &[u8] = &b;
+        let bytes: &[u8] = &b;
+        let kind = ContainerType::try_from_u8(bytes[0]).unwrap();
+        let mut bytes = &bytes[1..];
+        let depth = leb128::read::unsigned(&mut bytes).unwrap();
+        let (parent, bytes) = postcard::take_from_bytes(bytes).unwrap();
+        // SAFETY: bytes is a slice of b
+        let size = unsafe { bytes.as_ptr().offset_from(src.as_ptr()) };
```

Comment:
Is it not equal `src.len() - bytes.len()`?

Reply:
You're right!

Reply:
Fixed